src/main/scala/cognite/spark/v1/FilesRelation.scala

```diff
@@ -1,35 +1,144 @@
-      val file = fromRow[File](r)
-      file.copy(metadata = CdpConnector.filterMetadata(file.metadata))
-    }
-    client.files.updateFromRead(files) *> IO.unit
+    val files = fromRowWithFilteredMetadata(rows)
+    createOrUpdateByExternalId[File, FileUpdate, FileCreate, Files[IO]](
+      Set.empty,
+      files,
+      client.files,
+      doUpsert)
   }
 
+
+  def fromRowWithFilteredMetadata(rows: Seq[Row]): Seq[FileCreate] =
+    rows.map { r =>
+      val file = fromRow[FileCreate](r)
+      file.copy(metadata = filterMetadata(file.metadata))
+    }
+
   override def getStreams(filters: Array[Filter])(
       client: GenericClient[IO, Nothing],
       limit: Option[Int],
       numPartitions: Int): Seq[Stream[IO, File]] =
     Seq(client.files.list(limit))
 
+  override def insert(rows: Seq[Row]): IO[Unit] = {
+    val files = fromRowWithFilteredMetadata(rows)
+    client.files
+      .create(files)
+      .flatTap(_ => incMetrics(itemsCreated, files.size)) *> IO.unit
+  }
+
+  private def isUpdateEmpty(u: FileUpdate): Boolean = u == FileUpdate()
+
+  override def update(rows: Seq[Row]): IO[Unit] = {
+    val fileUpdates = rows.map(r => fromRow[FilesUpsertSchema](r))
+    updateByIdOrExternalId[FilesUpsertSchema, FileUpdate, Files[IO], File](
+      fileUpdates,
+      client.files,
+      isUpdateEmpty
+    )
+  }
+
+  override def delete(rows: Seq[Row]): IO[Unit] = {
+    val deletes = rows.map(r => fromRow[DeleteItem](r))
+    val ids = deletes.map(_.id)
+    client.files
+      .deleteByIds(ids)
+      .flatTap(_ => incMetrics(itemsDeleted, ids.length))
+  }
+
+  override def upsert(rows: Seq[Row]): IO[Unit] = {
+    val files = rows.map { r =>
+      val file = fromRow[FilesInsertSchema](r)
+      file.copy(metadata = filterMetadata(file.metadata))
+    }
+
+    val (itemsToUpdate, itemsToCreate) = files.partition(r => r.id.exists(_ > 0))
+
+    genericUpsert[File, FilesUpsertSchema, FileCreate, FileUpdate, Files[IO]](
+      itemsToUpdate.map(_.transformInto[FilesUpsertSchema]),
+      itemsToCreate.map(_.transformInto[FileCreate]),
+      isUpdateEmpty,
+      client.files)
+  }
+
   override def schema: StructType = structType[File]
 
   override def toRow(t: File): Row = asRow(t)
 
   override def uniqueId(a: File): Long = a.id
+
+object FilesRelation extends UpsertSchema {
+  val upsertSchema: StructType = structType[FilesUpsertSchema]
+  val insertSchema: StructType = structType[FilesInsertSchema]
+  val readSchema: StructType = structType[FilesReadSchema]
+}
+
+final case class FilesUpsertSchema(
+    id: Option[Long] = None,
+    externalId: Option[String] = None,
+    source: Option[String] = None,
+    metadata: Option[Map[String, String]] = None,
+    assetIds: Option[Seq[Long]] = None,
+    dataSetId: Option[Long] = None,
+    mimeType: Option[String] = None,
+    sourceCreatedTime: Option[Instant] = None,
+    sourceModifiedTime: Option[Instant] = None,
+    securityCategories: Option[Seq[Long]] = None,
```

Comment:
Remove the trailing comma, it's not allowed in Scala 2.11 and scalafmt doesn't like it.